### DISCUSSION TITLE
1st KRATOS COMMUNITY MEETING

### DISCUSSION THREAD

**Discussion by maceligueta** [KratosMultiphysics/Kratos#8239] [General]
Dear friends and colleagues, 

The KRATOS project is becoming everyday much bigger with the progressive inclusion of new features, and capabilities. This allows us to make simulations which were simply impossible only a few years ago but it also means that its complexity is increasing very fast.
The KRATOS community is also growing accordingly and many new members have joined the project in the last years.
In this regard, while a huge effort has been made to structure the “technical aspects” of the KRATOS development with the constitution of many committees and teams who are supervising the programming aspect and technical issues, we need now to build up a forum to give voice to the whole community. We need to create a constructive dialog between all the members not only to know each other and to better understand what everyone is doing to strengthen our collaborations and boost the potential of KRATOS itself, but also to understand the needs and the problems that our community is facing. 

This is the reason why we are inviting all of you to the 

# **1st KRATOS COMMUNITY MEETING**

### https://meet.google.com/zmb-nkpj-jte 

•	The meeting will be a forum to discuss the matters of the whole Kratos Community

•	It should help bridging the gap between the different groups in the Kratos Community each of those with different interests and skills (students, PhDs, postdocs, professors, industrial partners, ….)

•	The meeting may include technical questions but they should not be the core of the meeting

## HOW DOES IT WORK? 
We will collect **YOUR** questions, queries, suggestions in the next days. We will collect all the information and process it in the following days to arrive at a set of questions and a list of possible people to be ready to answer. This “preprocess” will allow to optimize the meeting and to keep it short and concise.

### •	Post your questions, doubts, issues and suggestions as contributions to this discussion before 12th February 2021

•	Use **‘thumb up’ or upvoting** to support previously written questions. The most popular questions will have priority

•	Join the meeting to strengthen our community!

## AGENDA for 26th February 2021 at 9.00 (CET)

▪	Welcome, introduction and presentation of the selected questions (5 mins)

▪	Answers/comments by selected speakers to the posed questions (10 mins) 

▪	Open discussion by ‘raising hands’ system (15 mins)

@maceligueta @antonialarese @rolandwuechner

**Comment by mokhairymahmoud** [comment]
Actually i had several questions in my mind the first time i checked Kratos. I will try to summarize some of them here : 

1) Why do not we see Kratos core inside on of the commercial software available in the market even though i think the software has all the credential to be a good commercial software and can solve industrial problems?

**Reply by maceligueta** [reply]
It would be great if questions a written separately, so people can upvote independently. You are also free to leave them as a pack, if you prefer that.

**Reply by mokhairymahmoud** [reply]
Okay no problem i will separate them.

**Reply by mpentek** [reply]
With the BSD license anyone could do that, perhaps some already do. How many commercial software solutions do you know of who transparently say which exact libraries they are using? I would assume more use open source software internally than we would think of.

I would also think about the Matlab-Python comparison, which goes in a similar direction. Commercial (and typically closed-source) software have different targets and philosophies behind.

Suggestion for a topic -> briefly but openly address the current state of Kratos and other commercial software and companies (as long as possible) to get a better overview.

**Comment by maceligueta** [comment]
Tagging @KratosMultiphysics/all to trigger notifications

**Comment by mokhairymahmoud** [comment]
2) Do you think that Kratos can compete with one of the gigantic software out there like Abaqus or Ansys? and if so why would not the companies replace these expensive software with Kratos for example?

**Reply by mpentek** [reply]
If you are interested in how it competes, you can have a validation campaign specific to your problem to see.

As far as adoption, in my opinion companies are not only focused on performance and quality, but vast established companies are tied down various other aspects, such as counting on the same workflow to work all the time, over many years. Also, they typically want on-demand support, which Ansys and Abaqus can afford to have. The Kratos community is for now not focused and also legally not (yet) prepared for that in my opinion.

As far as the current contributors go -> perhaps we can set up a section where current works (theses, publications) can be seen, also tagging which specifically do include validation.

**Comment by mokhairymahmoud** [comment]
3) My last question is regarding the documentation, since i think that Kratos documentation is very minute with respect to the documentation of Abaqus for example even though i think the size of Kratos is partially comparable to the size of Abaqus.

**Reply by jginternational** [reply]
Let me attach to this questions a note: What should we do with the old documentation / website/ wikis ... ?

**Reply by mpentek** [reply]
Who is supposed to keep up a vast documentation? Being an open-source project I think most contributors are happy to keep the functionalities up. Also, almost any contributor can add their own example, documentation. 

Here commercial software have a strong market incentive and use a lot of funds exactly for this. The also are more conservative to new and large updates and their functionalities are moving at a different (typically vastly slower) rate.

I also think that the main focus should shift to update and maintenance of existing examples, how-tos, and how to encourage new contributions while maintaining quality.

**Reply by sunethwarna** [reply]
Would it be helpfull to add meaningfull error messeages with an hint on how to rectify the error? Most of the time this is hidden. And when users compile Kratos in "Release" mode, even some of the normal errors are also not shown (not the full debug errors), they end up showing `terminate called after throwing an instance of 'Kratos::Exception'`, and this is not usefull for the user at all. When I run it through gdb, then I can see the actual error message (this is with Kratos compiled in `Release`). It would be really helpful for the users to see the actual error message with a small hint on how to correct it. [ This is related to error messages and checks done in c++ ]

**Comment by AlejandroCornejo** [comment]
Dear all, this is a very interesting proposal. I'll be posting my questions/suggestions:

1. Which is the approach that we are going to choose in terms of **pre/post processing software** compatible with **Kratos**. It has been noticed that the use of GiD is not always optimal (specially in industry). Maybe it would be interesting to discuss it.

**Reply by rubenzorrilla** [reply]
In this regard I think it makes sense that:

- @philbucher briefly comments on the current status of the Salome GUI
- @roigcarlo presents the LiteKratos GUI

**Reply by philbucher** [reply]
I guess I can discuss both the Salome stuff as well as the Paraview postprocessing

**Reply by pasenau** [reply]
Regarding _"It has been noticed that the use of GiD is not always optimal"_ 
What is missing from the Kratos interface for GiD? Can you (all) elaborate on some features or use-cases?

**Reply by maceligueta** [reply]
I think @AlejandroCornejo did not mean that there are missing features in the general sense, but in more corner cases instead. From what I know, there is a number of situations in which GiD is not helpful o necessary:
- Too large meshes/geometries (1e8 or 1e9 elements)
- Formats exported from CAD which can be calculated directly by Kratos (STL, IGA?,...)
- New developer features which haven't been added to the GUI yet.

This does not mean that GiD can't be used in general. It can, in my opinion. Anyway, @AlejandroCornejo can explain when he's back.

**Reply by AlejandroCornejo** [reply]
I'm sorry @pasenau If I did not explained myself clearly. My point is exactly what @maceligueta says. In my case I'm satisfied with the current status of the Kratos+GiD and with the capabilities of the software. There is no missing capabilities in this sense.

My point is more related to cases in which industry wants to use their own software that they trust/know more than GiD. They may not be  _better_ than GiD but they definetly have more expertise with (Abaqus, etc...). For example we have had requests of generating the meshes and submodels with Abaqus.

Summarizing, The people from CIMNE I think we are quite happy using GiD (even though as @maceligueta says the latest features in Structural have not been added but I understand that there's not enough time for everything) but I wanted to be honest and point out that industry have their own approach when it comes to this issue.

**Reply by philbucher** [reply]
Just to add that now the Salome plugin is available under the KratosMultiphysics organization:
https://github.com/KratosMultiphysics/KratosSalomePlugin

**Comment by riccardotosi** [comment]
What is your opinion of Kratos with respect to other open source software (if you have any)? Why should one prefer Kratos over other open-source software?
From [this question in Linkedin](https://www.linkedin.com/posts/holger-marschall-62175683_innovation-technology-multiphysics-activity-6763765643754934272-LmkG) you have a list of some open-source software:
- Basilisk (http://basilisk.fr/)
- code_saturne (https://lnkd.in/dte92kG)
- Elmer (https://lnkd.in/deX9SJS)
- FEniCS (https://fenicsproject.org)
- Fluidity (https://lnkd.in/dm68gZb)
- Kratos (https://lnkd.in/d-a2ENK)
- Nektar++ (https://www.nektar.info)
- OpenFOAM (https://www.openfoam.com)
- Sparselizard (https://lnkd.in/dPjnJH9)
- SU2 (https://su2code.github.io)

**Comment by maceligueta** [comment]
Sometimes I wonder whether some method, tool, process, utility,... is already implemented in Kratos. Is there a way to find that out? 
If not documented or searchable, can I ask in a fast, shared chat? Or should I always write an Issue?

**Reply by maceligueta** [reply]
Slack or Gitter may work for fast, brief chats, if we go that way.

**Reply by rubenzorrilla** [reply]
Just mentioning that at some point we tried Slack and the idea didn't succeed at all... Maybe because the community wasn't that large and we continued using private messaging.

**Reply by mokhairymahmoud** [reply]
is it possible to join this slack community or it is private for some people ?

**Reply by philbucher** [reply]
I think anyone could join
But I think it has been months if years since the last messages were written there :sweat_smile:

**Comment by loumalouomega** [comment]
We can drop Cotire if we set minimum version of CMake to 3.16: https://cmake.org/cmake/help/latest/variable/CMAKE_UNITY_BUILD.html#variable:CMAKE_UNITY_BUILD

**Reply by philbucher** [reply]
see cotire docs:
~~~
The functionality provided by cotire has been superseded by features added to CMake 3.16. 
Support for pre-compiling and unity builds is now built into CMake. 
Thus, there will not be any further updates or support for this project.
~~~

**Reply by rubenzorrilla** [reply]
I think this should be discussed in a technical meeting.

**Reply by loumalouomega** [reply]
FYI the transition can be done easily, I am doing some experiments right now in other project

**Reply by salvalatorre** [reply]
> I think this should be discussed in a technical meeting.

Totally agree. I understand this meeting is addressed to users, not developers.

**Reply by maceligueta** [reply]
It is actually addressed to the whole community, but it should not mainly focus on technical aspects. Technical discussions usually are of interest for a reduced group of people. It might happen, though, that a technical question gets upvoted many times.

**Comment by philbucher** [comment]
Probably not for the first meeting but here are three more things I / we (@KratosMultiphysics/technical-committee) wanted to communicate for a while already:
- How the CI (continuous integration) works => the checks that have to be passed before a PR can be merged. In the same note also go into how the testing is set up and working
- The future of shared memory parallelization in Kratos: The transition from OpenMP to C++ Threads (see [here](https://github.com/KratosMultiphysics/Kratos/wiki/Parallel-Utilities))
- The future of C++ testing in Kratos

**Comment by jginternational** [comment]
I'm interested in CD (Continuous Delivery). I did for myself some scripts and GH Actions in that way but, is there any "official channel" to download the last compiled version of the master branch, for a specific platform (unix or win64)?
If there's not a current way to do that, Do you think that this feature can be useful for the rest of the community?
Having the latest compiled zip instead of compiling by myself or using the release

**Comment by armingeiser** [comment]
Kratos as a python package (PyPi or conda) for easy delivery to project partners and easy access for first time users. What is the state/plan for this?

**Reply by loumalouomega** [reply]
See my comment below

**Reply by armingeiser** [reply]
I did create a custom one for a project partner, worked like a charm. But my question was more about regularly updated public ones, for all the default apps, hosted on PyPi for example.

**Comment by AlejandroCornejo** [comment]
Another useful point could be about the encapsulation of Kratos in an **executable file** to be delivered easily to early users/students withour having to compile Kratos inside each PC which can be costly and time consuming

**Reply by rubenzorrilla** [reply]
I think that this is pretty much the same as @armingeiser posted above.

**Reply by AlejandroCornejo** [reply]
> 
> 
> I think that this is pretty much the same as @armingeiser posted above.

Maybe yes... I'm talking also about generating our own `Kratos.exe` even with a custom implementation for a ad-hoc case etc

**Reply by rubenzorrilla** [reply]
Ah, you were pointing to a custom Kratos to be delivered. I think this can be now achieved with the Python wheels. Maybe it's worth it to discuss then.

**Reply by AlejandroCornejo** [reply]
yeah! imagine that you are testing some new capabilities with an student with a custom build in a research branch... Could be useful to generate Kratos.exe easily to be delivered to the user

**Reply by loumalouomega** [reply]
You can create custom pips and private ones (you require a server for that). This can be done easily in Gitlab (we can create a mirror just for this)

**Reply by pasenau** [reply]
Seems similar to what @jginternational proposed above about Continous Delivery (CD) and @armingeiser :
find a way to 'automatically' (triggered by some ?tag? and after passing the CI/CD tests) create an executable/zip/PyPi package at the same time, from the master or release branch.

**Comment by rubenzorrilla** [comment]
@maceligueta @antonialarese @rolandwuechner would it be possible to collect which are the most voted community queries to be discussed on Friday meeting? I think it would be helpful to a priori know which would be the topics to discuss about, specially for people expected to clarify things.

Many thanks in advance!

**Reply by maceligueta** [reply]
Only the three most popular topics will be addressed in this meeting, and the rest will be treated in the next one. Briefly:

- Python packaging
- Pre and post processing
- Kratos documentation and how to find stuff in Kratos 

We are keeping track of the rest of the topics, and will be ported to the next online discussion for the next meeting.

**Reply by antonialarese** [reply]
Thanks @rubenzorrilla, it is a good idea to publish the topics in advance. Just to mention that the people supposed to give an "answer"/comment were already informed some time ago.

**Comment by maceligueta** [comment]
The 1st Kratos Community Meeting took place this morning as planned. Many thanks to all the participants, especially to the speakers, on behalf of @antonialarese @rolandwuechner and myself.
The hottest topic was 'Documentation', that is why a new discussion has been opened ([Documentation](https://github.com/KratosMultiphysics/Kratos/discussions/8377)), but feel free to engage other meetings about the rest of the topics if you find it interesting.
As soon as we have all permissions from the participants appearing in the recording of the meeting we will upload it here.

**Reply by maceligueta** [reply]
Link to the recording: https://drive.google.com/file/d/1hZ4rmd6eJeOYdreGaPt94Nu7O5TMCSIH/view?usp=sharing

**Reply by jginternational** [reply]
Will it be available on Youtube? 
https://www.youtube.com/channel/UCzOncdymUoeudALah8JXnKA

**Reply by maceligueta** [reply]
If I was a Youtube user watching Kratos videos I would be surprised if the next video recommended is a meeting.
